### PR TITLE
Switch to NuGet packages for references.

### PR DESCRIPTION
**Pull request by KirillOsenkov**

### CODE REVIEW

**Comment by Krinkle** [comment]
@KirillOsenkov Aha, so you mean that with this one dependency added, the project works fine as-is on dotnet, without Mono? That makes sense historically, it just hadn't ocurred to me that there would not be anything else Mono-specific about CVNBot, or indeed that its built-ins would be readilly available on NuGet for use outside Mono. That's cool!

**Comment by Krinkle** [comment]
I've created a branch that adds `dotnet` as a test target:
* <https://github.com/countervandalism/CVNBot/compare/support-dotnet>

However this fails currently, not on absence of Sqlite, but on absence of .NET Framework:

<https://travis-ci.com/github/countervandalism/CVNBot/jobs/347363011>

Searching around, I get the sense that .NET Framework is only for Windows, and not available for via NuGet or something like that for dotnet on Linux or Mac. Is that correct?

I'd prefer for this to be verified/integrated without Mono, but I suppose that could be done later.

Or perhaps we could even drop .NET Framework as dependency. I'm actually not sure how to know whether we are currently using anything that isn't part of .NET Core/Standard.

File: src/CVNBot/CVNBot.csproj [hunk 1]
```diff
@@ -20,26 +20,20 @@
     <ExternalConsole>false</ExternalConsole>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="log4net, Version=2.0.8, Culture=neutral">
-      <SpecificVersion>False</SpecificVersion>
-      <HintPath>..\log4net.dll</HintPath>
-    </Reference>
-    <Reference Include="Meebey.SmartIrc4net, Version=1.1, Culture=neutral">
```
Comment (Krinkle) [inline src/CVNBot/CVNBot.csproj:27]: When I inspect the current embedded DLL for SmartIrc4net, VS2019 says it is actually version 0.4.5.0. But I'm not sure if I'm using interpreting this correctl and/or whether that is the one msbuild is indeed using.

**Comment by Krinkle** [comment]
If I understand correctly we can make it so that `msbuild /r` will install new dependencies (e.g. when adding or changing which packages are used), and that in general `msbuild` will only consider those already in the current repository.

I've spent a while trying to figure out how to tell msbuild (or NuGet) to store things in a specific directoy (e.g. `./packages`) and to also know to read from there at build time.

I got the latter to work by using the following in a src/CVNBot/NuGet.Config file:
```xml
    <packageRestore>
        <add key="enabled" value="False" />
        <add key="automatic" value="False" />
    </packageRestore>
```

However, I'm not sure how to make an explicit `msbuild /r` invocation still fetch it and put it somewhere for subsequent invocations to pick up.

File: src/CVNBot/CVNBot.csproj (hunk 1 above)
Comment (KirillOsenkov) [reply src/CVNBot/CVNBot.csproj:27]: This is correct. If you download the version 1.1 of the NuGet package and peek inside the .dll you'll notice that the version is 0.4.5.0 and the AssemblyInformationalVersion is 1.1.
image

**Comment by KirillOsenkov** [comment]
Honestly if you prefer to check in the .dlls I'd not use PackageReference at all then. Sounds like you already have it configured such that it works well when using Mono. For .NET Core I'd just also check in sqlite3.dll and Mono.Data.SQlite.dll. Trying to get a mix of two worlds is uncharted territory and I wouldn't recommend it. Either check in everything or use NuGet and its defaults.

Now to clear up some unfortunate confusion about the various flavors of .NET.

.NET Framework is Windows-only, legacy, "original" runtime, it goes up to 4.8 and 4.8 is the last major version. It installs with Windows and will be supported for a long time. It supports .NETStandard 2.0 and below.

Mono is the runtime created by Xamarin, it works on all OSs and it also supports .NETStandard 2.0 and as a bonus most libraries targeting .NET Framework also run on Mono.

.NET Core is the future evolution of .NET, it works on all OSs, is being actively funded, developed and is the safest bet for investments. Starting with .NET 5 we will drop the version "5" and just call it .NET going forward. It will include what we now know as .NET Core and it will also include the Mono runtime in case you want to run on devices where only Mono can run. They will share all the libraries.

I wouldn't recommend bothering with .NET Framework, but supporting .NET Core/.NET going forward would be a plus.

Regarding configuring NuGet to restore into a local directory such as packages is honestly a bit of uncharted territory. The strong recommendation is to just use the default conventions, PackageReferences, and while it is all extremely customizable you're on your own there, as we're trying to push the ecosystem into a common set of conventions/standards adopted by everyone.

Hope this helps!

**Comment by KirillOsenkov** [comment]
Oh and I just looked at the failure here: 
https://travis-ci.com/github/countervandalism/CVNBot/jobs/347363011

It complains about the reference assemblies for .NET Framework 4.7.2 not being installed. 

There are two ways to deal with this (if this was Windows you could just download the reference assemblies here: https://docs.microsoft.com/en-us/dotnet/framework/install/guide-for-developers)

1. switch from targeting `net472` to `netcoreapp3.1` here: 
https://github.com/countervandalism/CVNBot/blob/db7dbfd3f7237d28ee12b877230c68efcf237af2/src/CVNBot/CVNBot.csproj#L4
By targeting net472 you're targeting .NET Framework desktop (legacy), and while Mono supports this scenario out of the box so you don't have to do anything, with .NET Core it won't work if the reference assemblies aren't installed. By switching to `netcoreapp3.1` you're explicitly declaring that you're targeting .NET Core and it won't need anything else other than .NET Core on the machine that you can get from https://dotnet.microsoft.com/download/dotnet-core/3.1

2. perhaps an easier way is to just include the reference assemblies for net472 as a NuGet package, to make the build happy. You can do this by adding this line into an ItemGroup in the .csproj:

```
<PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0" />
```

**Pull request closed without merging**